### DISCUSSION TITLE
Security issue: Function Search Path Mutable

### DISCUSSION THREAD

**Discussion by VValiente** [supabase/supabase#23170] [Questions]
I want to insert a new row in the `user_profile` table for a new user after they sign up and I found this YouTube video [User Profile Creation on App Sign-up - Full walkthrough](https://www.youtube.com/hashtag/supabase). I followed the instructions to create the function and created a trigger via the SQL Editor as suggested by the top commentor.

Here's the SQL editor code:
For the function:
```
BEGIN
  INSERT INTO public.user_profile (id)
  VALUES (NEW.id);
  RETURN NEW;
END;
```

For the trigger:
```
create trigger new_user_trigger
after insert on auth.users for each row
execute function create_profile_for_new_users ();
```

I thought it worked because whenever a user is created, it adds a row in the correct table. Cascade delete also worked.

However, after watching Supabase's video about [8 things you should NEVER do in a Database!](https://www.youtube.com/watch?v=NZEbVe47DfA) a day after, I found the security warnings regarding the function I created.

I tried to follow the documentation on [Function Search Path Mutable](https://supabase.com/docs/guides/database/database-linter?queryGroups=lint&lint=0011_function_search_path_mutable) but I'm not able to update or drop the function due to some dependency. I suspect it is because the `auth` trigger I created via the SQL editor references it. I tried to drop that trigger but it won't let me even if it returns success. This is what the AI suggested: `drop trigger new_user_profile_trigger on auth.users;` Is this wrong? Or is it because the `auth` is managed privately? 

I don't know how to fix this security issue. Can someone please help? 

I'm a frontend developer and I'm new to this database thing.

**Comment by VValiente** [comment]
Came across this command to drop the function: `drop function create_user_profile_on_sign_up () cascade;` and that deleted both the function and the `auth` trigger.

I decided to perform an `insert` from the client on successful sign-up instead just to not get the warnings. It would've been nice to do this on the Supabase side.

If someone knows how this should be done, please share. Thanks in advance!

**Comment by Wasaanyi** [comment]
I have followed the guide from [https://supabase.com/docs/guides/database/database-advisors?queryGroups=lint&lint=0011_function_search_path_mutable](https://supabase.com/docs/guides/database/database-advisors?queryGroups=lint&lint=0011_function_search_path_mutable), 

for a given function like below
```sql
create function example_function()
  returns void
  language sql
as $$
  -- Your SQL code here
$$;
```

You can create or replace the function and add the search_path setting.
```sql
create or replace function example_function()
  returns void
  language sql
  set search_path = '' -- LOOK HERE
as $$
  -- Your SQL code here.
$$;
```

After doing that automatically on running the queries I was told I needed to access my tables as public.users instead of just users

**Reply by shimulroy842** [reply]
```sql
CREATE OR REPLACE FUNCTION fetch_initial_result(name_input text, type_input text)
RETURNS TABLE(endate date, bndate text, result_name text, result_type text, result_total bigint)
set search_path = ''
AS $$
BEGIN
    RETURN QUERY
select date as endate,bdate as bndate,name as result_name,type as result_type,sum(total)::int8 as result_total
  from main_sheet_b1430
  where name = name_input and
  type = type_input
  group by date,bdate,name,type
  order by endate DESC ;
END;
$$ LANGUAGE plpgsql;
```
the function creates successfully but when function executes it shows 

`ERROR:  42P01: relation "main_sheet_b1430" does not exist`

without `set search_path = ''` the function execute without error

**Reply by Wasaanyi** [reply]
Please try referencing the table using **public.main_sheet_b1430**

**Reply by shimulroy842** [reply]
using `public.main_sheet_b1430` solve the problem.

**Comment by eurlexa** [comment]
I was afraid to reaplace or even drop the functions in matter exactly because of some erros like this one. 

Just altering the functions worked for me :
```sql
ALTER FUNCTION funcion_name(arg type, arg type, ...)
SET search_path = ''; 
```

**Comment by atowersc** [comment]
In my case. 

```
CREATE OR REPLACE FUNCTION "public"."handle_new_user"() RETURNS "trigger"
    LANGUAGE "plpgsql"
    SET "search_path" TO ''
    AS $$
BEGIN
    INSERT INTO public.users (id, name, avatar_url)
    VALUES (
        new.id,
        new.raw_user_meta_data->>'full_name',
        new.raw_user_meta_data->>'avatar_url'
        );
    RETURN NEW;
END;
$$;
```

`SET "search_path" TO ''` removed the warning. but the vulnerability still persists. or not?

**Comment by apexskier** [comment]
I've hit this a couple times, and I think it's pretty easily addressed by the supabase team - the UI's function editor should add `SET "search_path" TO '' ` (or whatever the correct modification is) to functions created in the function editor so it's security by default. As is, it feels like you're guaranteed to have a warning if you use this feature.

**Comment by solvionsolutions** [comment]
Since your trigger on `auth.users` depends on the function, that dependency is exactly why the plain `drop` kept fighting you — and the good news is you don't have to drop anything to clear the warning. You can pin the search path on the existing function in place:

```sql
alter function public.create_profile_for_new_users() set search_path = '';
```

The one gotcha: once the search path is empty, nothing resolves by bare name anymore, so open the function body and fully qualify every reference — `public.user_profile`, `auth.users`, and so on — otherwise it'll start throwing "relation does not exist" the next time it fires. If you'd rather rebuild it cleanly instead of altering in place, do the same thing at creation time with `security definer set search_path = ''` and the qualified names baked in from the start. Both approaches silence the 0011 linter; the `alter` route just spares you from touching the trigger, which is nice given the `auth` schema is Supabase-managed. And keeping profile creation server-side via the trigger is perfectly fine — you don't need to push it to the client just to make the linter happy, though that's a reasonable design choice too if you prefer it.

I wrote up the full reasoning on why the empty search_path matters here: https://www.guardlayer.io/blog/supabase-security-definer-search-path
